### PR TITLE
feat(streaming): enable consistent hash for hash agg

### LINKED ISSUE #3544: Enable consistent hash for HashAgg

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR enables consistent hash for hash agg. The assertions in `CellBasedTable` shows that we've correctly sharded data with hash dispatcher.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- #3456 
- #3457
- #3514
- Close #3544

### CODE REVIEW

File: src/storage/src/table/cell_based_table.rs [hunk 1]
```diff
@@ -129,21 +129,22 @@ impl<S: StateStore, SER: RowSerializer> CellBasedTableBase<S, SER, READ_ONLY> {
     /// set of `column_ids`. The output will only contains columns with the given ids in the same
     /// order.
     /// This is parameterized on cell based row serializer.
+    // TODO: allow specifying the distribution keys and vnodes.
```
Comment (BugenZhao) [inline src/storage/src/table/cell_based_table.rs:132]: `CellBasedTable` is directly constructed only in batch query executor or seq scan. We leave them untouched, while specifying vnodes and dist keys for streaming-purpose `StateTable` is implemented.

**Review by xx01cyx — APPROVED** [review]
LGTM!

File: src/storage/src/table/cell_based_table.rs [hunk 2]
```diff
@@ -190,40 +200,41 @@ impl<S: StateStore, SER: RowSerializer> From<CellBasedTableBase<S, SER, READ_WRI
...
-        };
+        let dist_key_in_pk_indices = dist_key_indices
+            .iter()
+            .map(|&di| {
+                pk_indices
+                    .iter()
+                    .position(|&pi| di == pi)
+                    .expect("distribution keys must be a subset of primary keys")
+            })
+            .collect_vec();
```
Comment (BowenXiao1999) [inline src/storage/src/table/cell_based_table.rs:237]: Add some comments that this is check of dis key in pk_indices? I remembered we have similar code in Materialize Executor, which use `HashSet::is_subset` .

File: src/storage/src/table/cell_based_table.rs [hunk 3]
```diff
@@ -190,40 +200,41 @@ impl<S: StateStore, SER: RowSerializer> From<CellBasedTableBase<S, SER, READ_WRI
 }
 
 impl<S: StateStore, SER: RowSerializer, const T: AccessType> CellBasedTableBase<S, SER, T> {
+    pub(super) fn fallback_vnodes() -> Arc<Bitmap> {
```
Comment (BowenXiao1999) [inline src/storage/src/table/cell_based_table.rs:203]: So this function should only be used by singleton executor in future?

**Review by BowenXiao1999 — APPROVED** [review]
LSTM!

File: src/storage/src/table/cell_based_table.rs (hunk 2 above)
Comment (BugenZhao) [reply src/storage/src/table/cell_based_table.rs:237]: This is also used for computing the indices of distribution keys in pk. You may find some explanations here.

https://github.com/singularity-data/risingwave/blob/d6ea0a49c1ea43d61eb4d4193ea8f31f7cf81ffb/src/storage/src/table/cell_based_table.rs#L103-L105

File: src/storage/src/table/cell_based_table.rs (hunk 3 above)
Comment (BugenZhao) [reply src/storage/src/table/cell_based_table.rs:203]: Yep. Besides, if we don't want to enable consistent hash for some executors, we can also use this fallback distribution to always encode and decode `0x00` in the key.

**Comment by BugenZhao** [comment]
@mergifyio requeue

**Pull request merged**